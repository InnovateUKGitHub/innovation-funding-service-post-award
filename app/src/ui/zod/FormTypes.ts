enum FormTypes {
  // MSP Document Share page
  ProjectLevelDelete = "projectLevelDelete",
  PartnerLevelDelete = "partnerLevelDelete",
  ProjectLevelUpload = "projectLevelUpload",

  // Claim page
  ClaimLevelDelete = "claimLevelDelete",
  ClaimLevelUpload = "claimLevelUpload",
  ClaimForecastSaveAndContinue = "claimForecastSaveAndContinue",
  ClaimForecastSaveAndQuit = "claimForecastSaveAndQuit",
  ClaimLineItemSaveAndDocuments = "claimLineItemSaveAndDocuments",
  ClaimLineItemSaveAndQuit = "claimLineItemSaveAndQuit",

  // Project Setup
  ProjectSetupBankDetails = "projectSetupBankDetails",
  ProjectSetupContactAssociate = "projectSetupContactAssociate",
  ProjectSetupForecast = "projectSetupForecast",

  // Forecast Tile
  ForecastTileForecast = "forecastTileForecast",

  // Claim Details page
  ClaimDetailLevelUpload = "claimDetailLevelUpload",
  ClaimDetailLevelDelete = "claimDetailLevelDelete",

  // Claim review page
  ClaimReviewLevelSaveAndContinue = "claimReviewSaveAndContinue",
  ClaimReviewLevelUpload = "claimReviewLevelUpload",
  ClaimReviewLevelDelete = "claimReviewLevelDelete",

  // Project Change Requests
  ProjectChangeRequestCreate = "projectChangeRequestCreate",
  ProjectChangeRequestUpdateTypes = "projectChangeRequestUpdateTypes",

  PcrLevelDelete = "pcrLevelDelete",
  PcrLevelUpload = "pcrLevelUpload",

  PcrFinancialVirementsSummary = "pcrFinancialVirementsSummary",
  PcrFinancialVirementsCostCategorySaveAndContinue = "pcrFinancialVirementsCostCategorySaveAndContinue",
  PcrReallocateCostsChangeRemainingGrant = "pcrReallocateCostsChangeRemainingGrant",

  PcrApproveNewSubcontractorStep = "pcrApproveNewSubcontractorDetails",
  PcrApproveNewSubcontractorSummary = "pcrApproveNewSubcontractorSummary",

  PcrAddPartnerCompaniesHouseStepSaveAndContinue = "pcrAddPartnerCompaniesHouseStepSaveAndContinue",
  PcrAddPartnerCompaniesHouseStepSaveAndQuit = "pcrAddPartnerCompaniesHouseStepSaveAndQuit",

  PcrChangeProjectScopeProposedPublicDescriptionStepSaveAndContinue = "pcrChangeProjectScopeProposedPublicDescriptionStepSaveAndContinue",
  PcrChangeProjectScopeProposedProjectSummaryStepSaveAndContinue = "pcrChangeProjectScopeProposedProjectSummaryStepSaveAndContinue",
  PcrChangeProjectScopeSummary = "pcrChangeProjectScopeSummary",

  PcrRenamePartnerStep = "pcrRenamePartnerStep",
  PcrRenamePartnerFilesStep = "pcrRenamePartnerFilesStep",
  PcrRenamePartnerSummary = "pcrRenamePartnerSummary",

  PcrPrepareReasoningStep = "pcrPrepareReasoningStep",
  PcrPrepareReasoningFilesStep = "pcrPrepareReasoningFilesStep",
  PcrPrepareReasoningSummary = "pcrPrepareReasoningSummary",

  LoanLevelUpload = "loanLevelUpload",
  LoanLevelDelete = "loanLevelDelete",
  LoanRequest = "loanRequest",
}

export { FormTypes };
