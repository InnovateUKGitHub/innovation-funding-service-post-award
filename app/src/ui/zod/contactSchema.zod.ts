import { z } from "zod";
import { dateValidation, pclIdValidation } from "./helperValidators/helperValidators.zod";
import { makeZodI18nMap } from "@shared/zodi18n";

const contactSchemaErrorMap = makeZodI18nMap({ keyPrefix: ["contact"] });

const contactDtoSchema = z.object({
  id: pclIdValidation,
  associateStartDate: dateValidation,
});

const multipleContactDtoSchema = z.object({
  contacts: z.array(contactDtoSchema),
});

type MultipleContactDtoSchemaType = typeof multipleContactDtoSchema;
type ContactDtoSchemaType = typeof contactDtoSchema;

export {
  contactDtoSchema,
  multipleContactDtoSchema,
  contactSchemaErrorMap,
  ContactDtoSchemaType,
  MultipleContactDtoSchemaType,
};
