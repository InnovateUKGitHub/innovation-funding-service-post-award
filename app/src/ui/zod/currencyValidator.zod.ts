import { parseCurrency, validCurrencyRegex } from "@framework/util/numberHelper";
import { isNil } from "lodash";
import { z, ZodIssueCode } from "zod";

const defaultMaxCurrency = 999_999_999_999;
const defaultMinCurrency = 0;

export const getGenericCurrencyValidation = ({
  label,
  max = defaultMaxCurrency,
  min = defaultMinCurrency,
  required = false,
}: {
  label: string;
  max?: number;
  min?: number;
  required?: boolean;
}) => {
  return z
    .string()
    .nullable()
    .superRefine((val, ctx) => {
      const currency = parseCurrency(val);
      const isEmpty = (typeof val === "string" && val.trim() === "") || isNil(val);
      if (required && isEmpty) {
        return ctx.addIssue({
          code: ZodIssueCode.custom,
          params: {
            label,
            generic: ["currency"],
            i18n: "errors.required",
          },
        });
      }

      if (!isEmpty) {
        if (/[\$€¥]+/.test(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.not_pounds",
            },
          });
        }
        // Check if the string can even be parsed
        if (isNaN(currency)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.not_a_number",
            },
          });
        }

        // Make sure our currency isn't so big as to break our server
        if (currency > max) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              count: max,
              generic: ["currency"],
              i18n: "errors.too_big",
            },
          });
        }

        if (currency < min) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              count: min,
              generic: ["currency"],
              i18n: "errors.too_small",
            },
          });
        }
        if (/\.\d\d\d/.test(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.two_decimal_places",
            },
          });
        }

        if (!validCurrencyRegex.test(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              label,
              generic: ["currency"],
              i18n: "errors.invalid_currency",
            },
          });
        }
      }
    })
    .transform(x => {
      if ((typeof x === "string" && x.trim() === "") || isNil(x)) {
        return null;
      }
      return parseCurrency(x);
    });
};
