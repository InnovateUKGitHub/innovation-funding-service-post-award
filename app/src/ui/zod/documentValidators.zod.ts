import {
  allowedClaimDocuments,
  allowedImpactManagementClaimDocuments,
  allowedProjectLevelDocuments,
  DocumentDescription,
  allowedPcrLevelDocuments,
} from "@framework/constants/documentDescription";
import { IAppOptions } from "@framework/types/IAppOptions";
import { z } from "zod";
import { FormTypes } from "./FormTypes";
import {
  projectIdValidation,
  emptyStringToUndefinedValidation,
  partnerIdValidation,
  getMultiFileValidation,
  periodIdValidation,
  pcrItemIdValidation,
  costCategoryIdValidation,
  pcrIdValidation,
} from "./helperValidators.zod";
import { ProjectDto } from "@framework/dtos/projectDto";
import { ImpactManagementParticipation } from "@framework/constants/competitionTypes";
import { makeZodI18nMap } from "@shared/zodi18n";

type ProjectLevelUploadSchemaType = ReturnType<typeof getProjectLevelUpload>;
const getProjectLevelUpload = (config: IAppOptions) =>
  z.object({
    form: z.literal(FormTypes.ProjectLevelUpload),
    projectId: projectIdValidation,
    partnerId: z.union([emptyStringToUndefinedValidation, partnerIdValidation]),
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x => allowedProjectLevelDocuments.includes(x))
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    files: getMultiFileValidation(config),
  });

type UploadBankStatementSchemaType = ReturnType<typeof getBankStatementUpload>;
const getBankStatementUpload = (config: IAppOptions) =>
  z.object({
    form: z.literal(FormTypes.ProjectLevelUpload),
    projectId: projectIdValidation,
    partnerId: z.union([emptyStringToUndefinedValidation, partnerIdValidation]),
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x => [DocumentDescription.BankStatement].includes(x))
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    files: getMultiFileValidation(config),
  });

interface ClaimLevelUploadSchemaExtraProps {
  config: IAppOptions;
  project: Pick<ProjectDto, "impactManagementParticipation">;
}

type ClaimLevelUploadSchemaType = ReturnType<typeof getClaimLevelUpload>;
const documentsErrorMap = makeZodI18nMap({ keyPrefix: ["documents"] });
const getClaimLevelUpload = ({ config, project }: ClaimLevelUploadSchemaExtraProps) =>
  z.object({
    form: z.union([z.literal(FormTypes.ClaimLevelUpload), z.literal(FormTypes.ClaimReviewLevelUpload)]),
    projectId: projectIdValidation,
    partnerId: partnerIdValidation,
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x =>
          (project.impactManagementParticipation === ImpactManagementParticipation.Yes
            ? allowedImpactManagementClaimDocuments
            : allowedClaimDocuments
          ).includes(x),
        )
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    periodId: periodIdValidation,
    files: getMultiFileValidation(config),
  });

type ClaimDetailLevelUploadSchemaType = ReturnType<typeof getClaimDetailLevelUpload>;
const getClaimDetailLevelUpload = ({ config, project }: ClaimLevelUploadSchemaExtraProps) =>
  z.object({
    form: z.literal(FormTypes.ClaimDetailLevelUpload),
    projectId: projectIdValidation,
    partnerId: partnerIdValidation,
    costCategoryId: costCategoryIdValidation,
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x =>
          (project.impactManagementParticipation === ImpactManagementParticipation.Yes
            ? allowedImpactManagementClaimDocuments
            : allowedClaimDocuments
          ).includes(x),
        )
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    periodId: periodIdValidation,
    files: getMultiFileValidation(config),
  });

const getPcrLevelUpload = ({ config }: { config: IAppOptions }) =>
  z.object({
    form: z.literal(FormTypes.PcrLevelUpload),
    projectId: projectIdValidation,
    projectChangeRequestIdOrItemId: pcrItemIdValidation,
    description: z.union([
      emptyStringToUndefinedValidation,
      z.coerce
        .number()
        .refine(x => allowedPcrLevelDocuments.includes(x))
        .optional()
        .transform(x => x as DocumentDescription),
    ]),
    files: getMultiFileValidation(config),
  });

type PcrLevelUploadSchemaType = ReturnType<typeof getPcrLevelUpload>;

const projectLevelDelete = z.object({
  form: z.literal(FormTypes.ProjectLevelDelete),
  projectId: projectIdValidation,
  documentId: z.string(),
});

const partnerLevelDelete = z.object({
  form: z.literal(FormTypes.PartnerLevelDelete),
  projectId: projectIdValidation,
  partnerId: partnerIdValidation,
  documentId: z.string(),
});

type ClaimLevelDeleteSchemaType = typeof claimLevelDelete;
const claimLevelDelete = z.object({
  form: z.union([z.literal(FormTypes.ClaimLevelDelete), z.literal(FormTypes.ClaimReviewLevelDelete)]),
  projectId: projectIdValidation,
  partnerId: partnerIdValidation,
  periodId: periodIdValidation,
  documentId: z.string(),
});

const claimDetailLevelDelete = z.object({
  form: z.literal(FormTypes.ClaimDetailLevelDelete),
  projectId: projectIdValidation,
  partnerId: partnerIdValidation,
  periodId: periodIdValidation,
  costCategoryId: costCategoryIdValidation,
  documentId: z.string(),
});

const pcrLevelDelete = z.object({
  form: z.literal(FormTypes.PcrLevelDelete),
  projectId: projectIdValidation,
  projectChangeRequestIdOrItemId: z.union([pcrItemIdValidation, pcrIdValidation]),
  documentId: z.string(),
});

const projectOrPartnerLevelDelete = z.discriminatedUnion("form", [partnerLevelDelete, projectLevelDelete]);

export {
  getPcrLevelUpload,
  getProjectLevelUpload,
  projectLevelDelete,
  partnerLevelDelete,
  getClaimLevelUpload,
  getClaimDetailLevelUpload,
  claimDetailLevelDelete,
  claimLevelDelete,
  pcrLevelDelete,
  projectOrPartnerLevelDelete,
  getBankStatementUpload,
  documentsErrorMap,
};
export type {
  ProjectLevelUploadSchemaType,
  ClaimLevelUploadSchemaType,
  UploadBankStatementSchemaType,
  PcrLevelUploadSchemaType,
  ClaimDetailLevelUploadSchemaType,
  ClaimLevelDeleteSchemaType,
};
