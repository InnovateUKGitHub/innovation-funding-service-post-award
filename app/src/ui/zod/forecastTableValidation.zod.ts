import { makeZodI18nMap } from "@shared/zodi18n";
import {
  ClaimStatusGroup,
  getClaimStatusGroup,
} from "@ui/components/atomicDesign/organisms/forecasts/ForecastTable/getForecastHeaderContent";
import {
  mapToForecastTableDto,
  MapToForecastTableProps,
} from "@ui/components/atomicDesign/organisms/forecasts/ForecastTable/NewForecastTable.logic";
import { z, ZodIssueCode } from "zod";
import { FormTypes } from "./FormTypes";
import {
  booleanValidation,
  partnerIdValidation,
  profileIdValidation,
  projectIdValidation,
} from "./helperValidators.zod";
import { getGenericCurrencyValidation } from "./currencyValidator.zod";

type ForecastTableSchemaType = ReturnType<typeof getForecastTableValidation>["schema"];
const getForecastTableValidation = (data: Omit<MapToForecastTableProps, "clientProfiles">) => {
  const errorMap = makeZodI18nMap({ keyPrefix: ["forecastTable"], context: data });

  const schema = z
    .object({
      projectId: projectIdValidation,
      partnerId: partnerIdValidation,
      form: z.union([
        z.literal(FormTypes.ClaimForecastSaveAndContinue),
        z.literal(FormTypes.ClaimForecastSaveAndQuit),
        z.literal(FormTypes.ProjectSetupForecast),
        z.literal(FormTypes.ForecastTileForecast),
      ]),
      profile: z.record(
        profileIdValidation,
        getGenericCurrencyValidation({
          label: "forms.forecastTable.profile.label",
          min: -1_000_000_000,
          required: true,
        }),
      ),
      submit: booleanValidation,
    })
    .superRefine(({ profile: clientProfiles, form, submit }, { addIssue, path }) => {
      const table = mapToForecastTableDto({ ...data, clientProfiles });
      const nonForecastClaims = data.claimTotalProjectPeriods.filter(
        x => getClaimStatusGroup(x.status) !== ClaimStatusGroup.FORECAST,
      );
      const finalClaim = nonForecastClaims.find(claim => claim.isFinalClaim);

      if (form === FormTypes.ProjectSetupForecast) {
        if (submit) {
          for (const costCategory of table.costCategories) {
            if (costCategory.differentThanAllocatedCosts && !costCategory.isCalculated) {
              addIssue({
                code: ZodIssueCode.custom,
                path: [...path, "costCategory", costCategory.costCategoryId],
                params: {
                  costCategoryName: costCategory.costCategoryName,
                  i18n: "errors.forecasts_different",
                },
              });
            }
          }
        }
      } else {
        // Don't check if the total is too big for project setup spend profile
        // (there is already a check for each cost cat)
        if (table.totalRow.total > table.totalRow.golCost) {
          addIssue({
            code: ZodIssueCode.too_big,
            maximum: table.totalRow.golCost,
            inclusive: false,
            type: "number",
            path: [...path, "total"],
          });
        }
      }

      if (clientProfiles && finalClaim) {
        addIssue({
          code: ZodIssueCode.custom,
          params: {
            i18n: finalClaim.isApproved ? "errors.final_claim_already_approved" : "errors.final_claim",
          },
        });
      }
    });

  return { errorMap, schema };
};

export { getForecastTableValidation };
export type { ForecastTableSchemaType };
