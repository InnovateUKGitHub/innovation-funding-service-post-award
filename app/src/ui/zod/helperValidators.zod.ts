import { SalesforcePrefixes } from "@framework/constants/salesforceConstants";
import { ClientFileWrapper } from "@client/clientFileWrapper";
import { z, ZodIssueCode } from "zod";
import { IsomorphicFileWrapper } from "@server/apis/isomorphicFileWrapper";
import { validDocumentFilenameCharacters } from "@ui/validation/validators/documentUploadValidator";
import { getFileExtension, getFileName } from "@framework/util/files";
import { IAppOptions } from "@framework/types/IAppOptions";
import { IFileWrapper } from "@framework/types/fileWapper";

const projectIdValidation = z
  .string()
  .startsWith(SalesforcePrefixes.Acc_Project__c)
  .transform(x => x as ProjectId);

const pcrIdValidation = z
  .string()
  .startsWith(SalesforcePrefixes.Acc_ProjectChangeRequest__c)
  .transform(x => x as PcrId);

const pcrItemIdValidation = z
  .string()
  .startsWith(SalesforcePrefixes.Acc_ProjectChangeRequest__c)
  .transform(x => x as PcrItemId);

const partnerIdValidation = z
  .string()
  .startsWith(SalesforcePrefixes.Acc_ProjectParticipant__c)
  .transform(x => x as PartnerId);

const emptyStringToUndefinedValidation = z
  .string()
  .refine(x => x === "")
  .transform(() => undefined)
  .optional();

const emptyStringToNullValidation = z
  .string()
  .refine(x => x === "")
  .transform(() => null)
  .nullable();

const periodIdValidation = z.coerce
  .number()
  .int()
  .gt(0)
  .lt(500) // Assumption that a project has fewer than 500 periods.
  .transform(x => x as PeriodId);

const getSingleFileValidation = (options: IAppOptions) => {
  const { imageTypes, pdfTypes, presentationTypes, spreadsheetTypes, textTypes } = options.permittedTypes;
  const permittedFileTypes = [...pdfTypes, ...textTypes, ...presentationTypes, ...spreadsheetTypes, ...imageTypes];

  return z.custom<IFileWrapper>().superRefine((file, ctx) => {
    const basename = getFileName(file.fileName);
    const extension = getFileExtension(file.fileName);

    if (basename.length === 0 || extension.length === 0) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_basename_too_small",
        },
      });
    }

    if (!(file.fileName.length <= options.maxFileBasenameLength)) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_basename_too_big",
          count: options.maxFileBasenameLength,
        },
      });
    }

    if (!validDocumentFilenameCharacters.test(file.fileName)) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_name_invalid_characters",
        },
      });
    }

    if (!permittedFileTypes.includes(extension)) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_extension_invalid_type",
        },
      });
    }

    if (!(file.size > 0)) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_size_too_small",
        },
      });
    }

    if (!(file.size <= options.maxFileSize)) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        params: {
          i18n: "errors.file_size_too_large",
          size: options.maxFileSize,
        },
      });
    }
  });
};

const getMultiFileValidation = (options: IAppOptions) =>
  z.preprocess((x: unknown) => {
    // Map to ClientFileWrapper/ServerFileWrapper
    if (Array.isArray(x) && x.every(x => x instanceof IsomorphicFileWrapper)) return x;
    if ("FileList" in globalThis && x instanceof FileList) return [...x].map(x => new ClientFileWrapper(x));
    return null;
  }, z.array(getSingleFileValidation(options)).min(1).max(options.maxUploadFileCount));

export {
  projectIdValidation,
  pcrIdValidation,
  pcrItemIdValidation,
  partnerIdValidation,
  periodIdValidation,
  emptyStringToUndefinedValidation,
  emptyStringToNullValidation,
  getSingleFileValidation,
  getMultiFileValidation,
};
