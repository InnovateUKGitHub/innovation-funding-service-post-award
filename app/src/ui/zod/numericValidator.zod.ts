import { isNil } from "lodash";
import { z, ZodIssueCode } from "zod";

const defaultMax = 999_999_999_999;
const defaultMin = 0;

/**
 * if `required` return type is `"number"`
 * if not `required` return type is `"null | number"`
 */
type NullableIfNotRequired<Required extends boolean> = Required extends true
  ? z.ZodEffects<z.ZodEffects<z.ZodNumber>>
  : // eslint-disable-next-line @typescript-eslint/no-explicit-any
    z.ZodEffects<z.ZodEffects<z.ZodOptional<z.ZodNullable<z.ZodAny>>, any, any>, number | null, any>;

/**
 * ### getNumericValidation
 *
 * this will match a set of validations and
 * also error messaging to ensure consistency across the app
 * and compliance with GDS and is for evaluating numeric values
 *
 */
export const getNumberValidation = <T extends boolean>({
  max = defaultMax,
  min = defaultMin,
  lt, // less than
  gt, // greater than
  required,
  integer, // will reject if input not an integer
  showValidRange, // will cause the error to show both min and max values
  decimalPlaces, // will have an error that shows the maximum allowed decimal places. Only works with string type values
}: {
  max?: number;
  min?: number;
  lt?: number;
  gt?: number;
  required: T;
  integer?: boolean;
  showValidRange?: boolean;
  decimalPlaces?: number;
}): NullableIfNotRequired<T> =>
  z
    .any()
    .nullable()
    .optional()
    .superRefine((rawVal, ctx) => {
      /*
       * run the required check first
       */
      const isEmpty = isNil(rawVal);
      if (required && isEmpty) {
        return ctx.addIssue({
          code: ZodIssueCode.custom,
          params: {
            generic: true,
            i18n: "errors.generic.number.required",
          },
        });
      }

      /*
       * run the rest of the checks
       */
      if (!isEmpty) {
        /*
         * if we have a decimal places check, run this. This only makes sense on string types
         */
        if (typeof rawVal === "string" && !isNil(decimalPlaces)) {
          const decimalPlacesRegex = /\d+\.(\d*)?/;
          const match = rawVal.match(decimalPlacesRegex);
          if (match) {
            const decimalPlacesMatch = match[1];
            if (decimalPlacesMatch?.length > decimalPlaces) {
              return ctx.addIssue({
                code: ZodIssueCode.custom,
                params: {
                  generic: true,
                  i18n: "errors.generic.number.decimal_places",
                  count: decimalPlaces,
                },
              });
            }
          }
        }

        /*
         * convert the value to number type
         */
        let val: number;

        if (typeof rawVal === "number") {
          val = rawVal;
        } else {
          val = Number(rawVal);
        }

        /*
         * reject if not a number
         */
        if (isNaN(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              generic: true,
              i18n: "errors.generic.number.invalid_type",
            },
          });
        }

        /*
         * check for integer if this is required
         */
        if (integer && !Number.isInteger(val)) {
          return ctx.addIssue({
            code: ZodIssueCode.custom,
            params: {
              generic: true,
              i18n: "errors.generic.number.not_integer",
            },
          });
        }

        /*
         * if show valid range is passed in then the error message shows both
         * min and max allowed values
         */
        if (showValidRange) {
          if (val < min || val > max) {
            return ctx.addIssue({
              code: ZodIssueCode.custom,
              params: {
                min,
                max,
                count: min === 1 && max === 1 ? 1 : 2,
                generic: true,
                i18n: "errors.generic.number.invalid_range",
              },
            });
          }
        } else {
          /*
           * reject if too big
           */
          if (val > max) {
            return ctx.addIssue({
              code: ZodIssueCode.custom,
              params: {
                count: max,
                generic: true,
                i18n: "errors.generic.number.too_big",
              },
            });
          }

          /*
           * also check for less than
           */
          if (typeof lt === "number" && val >= lt) {
            return ctx.addIssue({
              code: ZodIssueCode.custom,
              params: {
                count: lt,
                generic: true,
                i18n: "errors.generic.number.less_than",
              },
            });
          }

          /*
           * check for too small
           */
          if (val < min) {
            return ctx.addIssue({
              code: ZodIssueCode.custom,
              params: {
                count: min,
                generic: true,
                i18n: "errors.generic.number.too_small",
              },
            });
          }

          /*
           * check for greater than
           */
          if (typeof gt === "number" && val <= gt) {
            return ctx.addIssue({
              code: ZodIssueCode.custom,
              params: {
                count: lt,
                generic: true,
                i18n: "errors.generic.number.greater_than",
              },
            });
          }
        }
      }
    })
    .transform(val => {
      /*
       * transform result into number type or null
       */
      if (!required && isNil(val)) {
        return null;
      } else if (typeof val === "number") {
        return val;
      } else {
        return Number(val);
      }
    }) as NullableIfNotRequired<T>;
