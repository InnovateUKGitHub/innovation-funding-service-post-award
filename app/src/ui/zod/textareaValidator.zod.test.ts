import { initStubTestIntl } from "@shared/initStubTestIntl";
import { getTextareaValidation } from "./textareaValidator.zod";
import { makeZodI18nMap } from "@shared/zodi18n";

describe("generic textarea validator", () => {
  beforeAll(async () => {
    await initStubTestIntl({
      forms: {
        errors: {
          generic: {
            textarea: {
              required: "{{ label }} required!",
              invalid_range: "{{ label }} {{ count }} not between {{ min }} {{ max }}!",
              too_small: "{{ label }} must be bigger than {{ count }}!",
              too_big: "{{ label }} must be smaller than {{ count }}!",
            },
          },
        },
        my: {
          nice: {
            error: {
              label: "ARTMS",
            },
          },
        },
      },
    });
  });

  test.each([
    ["non required empty", null, false, true],
    ["required empty", null, true, false],
    ["too small", "ops", true, false],
    ["too big", "this is much too big and impossible", true, false],
    ["acceptable", "acceptable", true, true],
  ])("%s", (name, input, required, accept) => {
    const errorMap = makeZodI18nMap({ keyPrefix: ["my", "nice", "error"] });
    const parse = getTextareaValidation({
      maxLength: 20,
      minLength: 5,
      required: required,
    }).safeParse(input, { errorMap });
    expect(parse.success).toBe(accept);
    expect(parse).toMatchSnapshot();
  });
});
