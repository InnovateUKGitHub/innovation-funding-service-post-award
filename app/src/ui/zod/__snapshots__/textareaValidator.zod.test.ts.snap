// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generic textarea validator max length validation 3 chars long 1`] = `
{
  "data": {
    "organisationName": "ops",
  },
  "success": true,
}
`;

exports[`generic textarea validator max length validation 10 chars long 1`] = `
{
  "data": {
    "organisationName": "acceptable",
  },
  "success": true,
}
`;

exports[`generic textarea validator max length validation 35 chars long 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "count": 20,
      "generic": true,
      "i18n": "errors.generic.textarea.too_big"
    },
    "path": [
      "organisationName"
    ],
    "message": "Organisation name must be smaller than 20!"
  }
]],
  "success": false,
}
`;

exports[`generic textarea validator max length validation non required empty 1`] = `
{
  "data": {
    "organisationName": undefined,
  },
  "success": true,
}
`;

exports[`generic textarea validator max length validation required empty 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "generic": true,
      "i18n": "errors.generic.textarea.required"
    },
    "path": [
      "organisationName"
    ],
    "message": "Organisation name required!"
  }
]],
  "success": false,
}
`;

exports[`generic textarea validator max length validation spaces are not allowed 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "generic": true,
      "i18n": "errors.generic.textarea.required"
    },
    "path": [
      "organisationName"
    ],
    "message": "Organisation name required!"
  }
]],
  "success": false,
}
`;

exports[`generic textarea validator max length validation spaces are treated as undefined 1`] = `
{
  "data": {
    "organisationName": undefined,
  },
  "success": true,
}
`;

exports[`generic textarea validator ranged validation acceptable 1`] = `
{
  "data": "acceptable",
  "success": true,
}
`;

exports[`generic textarea validator ranged validation non required empty 1`] = `
{
  "data": undefined,
  "success": true,
}
`;

exports[`generic textarea validator ranged validation required empty 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "generic": true,
      "i18n": "errors.generic.textarea.required"
    },
    "path": [],
    "message": "ARTMS required!"
  }
]],
  "success": false,
}
`;

exports[`generic textarea validator ranged validation spaces are not allowed 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "generic": true,
      "i18n": "errors.generic.textarea.required"
    },
    "path": [],
    "message": "ARTMS required!"
  }
]],
  "success": false,
}
`;

exports[`generic textarea validator ranged validation spaces are treated as undefined 1`] = `
{
  "data": undefined,
  "success": true,
}
`;

exports[`generic textarea validator ranged validation too big 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "min": 5,
      "max": 20,
      "count": 35,
      "generic": true,
      "i18n": "errors.generic.textarea.invalid_range"
    },
    "path": [],
    "message": "ARTMS 35 not between 5 20!"
  }
]],
  "success": false,
}
`;

exports[`generic textarea validator ranged validation too small 1`] = `
{
  "error": [ZodError: [
  {
    "code": "custom",
    "params": {
      "min": 5,
      "max": 20,
      "count": 3,
      "generic": true,
      "i18n": "errors.generic.textarea.invalid_range"
    },
    "path": [],
    "message": "ARTMS 3 not between 5 20!"
  }
]],
  "success": false,
}
`;
